### PR TITLE
White spaces around inline object elements should not be lost in data.

### LINKED ISSUE #10147: Whitespace gets removed when adjacent to inline image
## 📝 Provide detailed reproduction steps (if any)

1. Open a demo, e.g. https://ckeditor.com/docs/ckeditor5/latest/examples/builds/classic-editor.html .
2. Execute in the console:
```js
editor.setData(
    '<p>foo <img src="../../assets/img/volcano.jpg"></p>'
)
```
or
```js
editor.setData(
    '<p><img src="../../assets/img/volcano.jpg"> bar</p>'
)
```

### ✔️ Expected result

Spaces after `foo` and before `bar` are preserved.

### ❌ Actual result

Spaces are removed.

### ❓ Further details

In the `DomConverter#_processDataFromDomText()` method there are calls for picking adjacent inline elements.

https://github.com/ckeditor/ckeditor5/blob/a39c5ed569c1f73436661ff494e4548ee54a8866/packages/ckeditor5-engine/src/view/domconverter.js#L1112-L1113

The problem is that method `DomConverter#_getTouchingInlineDomNode()` accepts only text and `<br>` as an inline element:

https://github.com/ckeditor/ckeditor5/blob/a39c5ed569c1f73436661ff494e4548ee54a8866/packages/ckeditor5-engine/src/view/domconverter.js#L1265-L1277

Next call in `_processDataFromDomText()` is `_checkShouldLeftTrimDomText()`. Which is only checking if a node is an element without checking if the element is an inline kind. This returns `true` for `imageInline` and allows the parent method to remove the space

https://github.com/ckeditor/ckeditor5/blob/a39c5ed569c1f73436661ff494e4548ee54a8866/packages/ckeditor5-engine/src/view/domconverter.js#L1168-L1175

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by oleq**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Fix (engine): White spaces around inline object elements such as `` or `<button>` should not be lost in data. Closes #10147.

Tests (image): White spaces around inline images should not be lost in data (see #10147).

Internal (markdown-gfm): Changed the way checkboxes are rendered by marked to compensate for the new way of handling white spaces around inline objects (see #10147).

Tests (paste-from-office): Enabled inline images in the paste from office tests for more realistic integration testing. Allowed execution of a single test among automatically generated ones (see #10147).

### CODE REVIEW

File: packages/ckeditor5-engine/src/view/domconverter.js [hunk 1]
```diff
@@ -1258,45 +1283,52 @@ export default class DomConverter {
...
+				return null;
+			}
+		} while (
+			!( isText( node ) || node.tagName == 'BR' || this._isInlineObjectElement( node ) )
+		);
 
-		if ( touchingNode !== null ) {
-			const lca = getCommonAncestor( node, touchingNode );
+		return node;
+	}
```
Comment (maxbarnas) [inline packages/ckeditor5-engine/src/view/domconverter.js:1310]: Ohhh, that's nice!

File: packages/ckeditor5-engine/src/view/domconverter.js [hunk 2]
```diff
@@ -1141,14 +1157,17 @@ export default class DomConverter {
 		// ` \u00A0` to ensure proper rendering. Since here we convert back, we recognize those pairs and change them back to `  `.
 		data = data.replace( / \u00A0/g, '  ' );
 
+		const isNextNodeInlineObjectElement = nextNode && this.isElement( nextNode ) && nextNode.tagName != 'BR';
+		const isNextNodeStaringWithSpace = nextNode && isText( nextNode ) && nextNode.data.charAt( 0 ) == ' ';
```
Comment (maxbarnas) [inline packages/ckeditor5-engine/src/view/domconverter.js:1161]:
```suggestion
		const isNextNodeStartingWithSpace = nextNode && isText( nextNode ) && nextNode.data.charAt( 0 ) == ' ';
```

File: packages/ckeditor5-engine/src/view/domconverter.js [hunk 3]
```diff
@@ -1141,14 +1157,17 @@ export default class DomConverter {
...
+
 		// Then, let's change the last nbsp to a space.
-		if ( /( |\u00A0)\u00A0$/.test( data ) || !nextNode || ( nextNode.data && nextNode.data.charAt( 0 ) == ' ' ) ) {
+		if ( /( |\u00A0)\u00A0$/.test( data ) || !nextNode || isNextNodeInlineObjectElement || isNextNodeStaringWithSpace ) {
```
Comment (maxbarnas) [inline packages/ckeditor5-engine/src/view/domconverter.js:1164]:
```suggestion
		if ( /( |\u00A0)\u00A0$/.test( data ) || !nextNode || isNextNodeInlineObjectElement || isNextNodeStartingWithSpace ) {
```

**Review by maxbarnas — APPROVED** [review]
LGTM 👯

File: packages/ckeditor5-engine/src/view/domconverter.js (hunk 1 above)
Comment (niegowski) [reply packages/ckeditor5-engine/src/view/domconverter.js:1310]: Thanks 😉

**Pull request merged**